odoo/orm/models.py

```diff
@@ -6292,28 +6295,21 @@ def _convert_to_write(self, values):
     # Record traversal and update
     #
 
-    def _mapped_func(self, func):
-        """ Apply function ``func`` on all records in ``self``, and return the
-            result as a list or a recordset (if ``func`` returns recordsets).
-        """
-        if self:
-            vals = [func(rec) for rec in self]
-            if isinstance(vals[0], BaseModel):
-                return vals[0].union(*vals)         # union of all recordsets
-            return vals
-        else:
-            vals = func(self)
-            return vals if isinstance(vals, BaseModel) else []
-
-    def mapped(self, func):
-        """Apply ``func`` on all records in ``self``, and return the result as a
-        list or a recordset (if ``func`` return recordsets). In the latter
-        case, the order of the returned recordset is arbitrary.
+    def mapped(self, func: str | Callable) -> list:
+        """Apply ``func`` on all records in ``self``, and return a list of
+        mapped results on each record.
 
         :param func: a function or a dot-separated sequence of field names
         :type func: callable or str
-        :return: self if func is falsy, result of func applied to all ``self`` records.
-        :rtype: list or recordset
+        :return: list of func applied to each record in ``self``
+        :rtype: list of mapped results (same size as self)
+
+        The function is roughly equivalent to a list comprehension in python:
+        ``list(map(func, self))``.
+        In case of a dot-separated sequence, we traverse the path for each
+        record to get a single value. When the last field is not relational
+        and we have traversed an x2many field in the path, we use ``mapped``
+        because we may have multiple values.
```

Comment:
in general I like the idea of always returing the same return type

but according to this maybe we shouldn't claim
`:rtype: list of mapped results (same size as self)`

aslo the implicit deduplicating logic is removed for many2one fieds in the dot-separated string, if not other x2many appears in the string. I am not sure if it is OK and if anyone uses the side-effect.
And since the behaviors for x2many and many2one in the dot-separated string are different, it may be more tricky to write the code. Especially when the str func comes from the client, you don't know if any x2many field is in it when you call `mapped`

can we be more conservative to return size is equal to the second-to-last field value?
e.g `a.mapped('b.c.d') == a['b']['c'].mapped('d')`

Reply:
In all cases when the input was taken from the user, I have rewritten the code so that we explicitly state what we want (mostly without using `mapped`).

The more conservative approach seems tempting, because it is closed to current behaviour. However, the use case for having the same size is to ask questions like "what are the name of partners for these records?". 
The conservative method is still not mapping records as it deduplicates entries and sorts using the comodel.

```python
# ** same size as self
names = records.mapped('partner_id.name')
# same as
list(map(lambda r: r.partner_id.name, records))
zip(records, names)  # works well

# ** conservative
names = records.mapped('partner_id.name')
assert len(names) <= len(records)  # and out of order
# you could have written this (and the script does that migration)
assert names == records.partner_id.mapped('name')
```

For the edge case (with x2many): if the last field is non-relational and in the path there was an x2many record, use `mapped` again because going through a x2many record no longer ensures there is a single value for a record, so we should return a list in that case.